### DISCUSSION TITLE
RecursionError in jsonable_encoder when trying to encode an object with circular reference

### DISCUSSION THREAD

**Discussion by Abdallah-Darwish** [fastapi/fastapi#10358] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi.encoders import jsonable_encoder

lst = [1, 2, 3]
lst.append(lst)

encoded_list = jsonable_encoder(lst)
```

### Description

When trying to encode an object with recursive reference you'll get a RecursionError instead of a more descriptive error like `json.dumps` "Circular reference detected" .

When using `json.dump` you get the option to check for circular references which would make your life easier as a developer, so it would be nice to have something similar in fastapi.

### Operating System

macOS

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

3.11.5

### Additional Context

[json.dump](https://docs.python.org/3/library/json.html#json.dump)

**Comment by Abdallah-Darwish** [comment]
This is my attempt at solving it https://github.com/tiangolo/fastapi/pull/10359 and making it similar to pythons standard behaviour

**Comment by Kludex** [comment, marked as answer]
Well... But it is a circular reference... 🤔 

I don't think there's anything wrong here.

**Reply by Abdallah-Darwish** [reply]
As stated in the description the issue is with the exception type not that there is no exception being thrown at all, please do check `json.dump`
